Add support for Ephemeral Containers to the kubelet

verb opened the pull request:
**What this PR does / why we need it**: This implements the [Troubleshooting Running Pods](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/troubleshoot-running-pods.md) in the kubelet, protected by a feature flag.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
WIP #10834
WIP #27140

**Special notes for your reviewer**:
/assign @yujuhong 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Ephemeral containers have been added in alpha. These temporary containers can be added to running pods for purposes such as debugging, similar to how `kubectl exec` runs an process in an existing container. Also like `kubectl exec`, no resources are reserved for ephemeral containers and they are not restarted when they exit.

Note that container namespace targeting is not yet implemented, so [process namespace sharing](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/) must be enabled to view process from other containers in the pod.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://git.k8s.io/enhancements/keps/sig-node/20190212-ephemeral-containers.md
```

verb commented:
/retest

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -354,9 +355,10 @@ type containerToKillInfo struct {
+// Note that podActions will not affect Debug Containers since they are not part of the spec.
 type podActions struct {
-	// Stop all running (regular and init) containers and the sandbox for the pod.
+	// Stop all running (regular, init and debug) containers and the sandbox for the pod.
```
This seems to conflict with the above comment stating that the actions don't affect debug containers.

tallclair commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -105,6 +106,14 @@ func newContainerLabels(container *v1.Container, pod *v1.Pod, containerType kube
 		labels[types.KubernetesContainerTypeLabel] = string(containerType)
+		// It might be useful to store the spec for every container, but for now only store spec for debug containers
+		if containerType == kubecontainer.ContainerTypeEphemeral {
+			if containerJSON, e := json.Marshal(container); e != nil {
```
Does size matter here? If it does a base64 encoded protobuf will probably be more efficient.

tallclair commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -105,6 +106,14 @@ func newContainerLabels(container *v1.Container, pod *v1.Pod, containerType kube
+		// It might be useful to store the spec for every container, but for now only store spec for debug containers
+		if containerType == kubecontainer.ContainerTypeEphemeral {
+			if containerJSON, e := json.Marshal(container); e != nil {
+				glog.Errorf("Unable to marshal %v container %q of pod %v: %v", containerType, container, pod.Name, e)
```
nit: use [`format.Pod(pod)`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/util/format/pod.go#L32) in place of the pod name.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -809,6 +834,30 @@ func (m *kubeGenericRuntimeManager) RunInContainer(id kubecontainer.ContainerID,
+		return errors.New("Debug Containers feature disabled")
+	}
+
+	if kubecontainer.GetContainerSpec(pod, container.Name) != nil {
```
What if there's another ephemeral container with the same name? Is there a race condition here?

verb commented:
I've updated this PR to reflect updates to the API from kubernetes/community#1269, which makes the kubelet change simpler.

Note that tests will fail until this PR is rebased to include #59416.

verb commented:
/remove-lifecycle stale

spiffxp commented:
ping, needs rebase

spiffxp commented:
/milestone v1.13
I'm adding this to the v1.13 milestone since it relates to https://github.com/kubernetes/features/issues/277 which is currently planned for v1.13

guineveresaenger commented:
@verb given that this PR still is a WIP, what is your level of confidence this can be ready for the 1.13 milestone? Enhancement freeze is tomorrow end of day. If there is no communication or update on the PR, this is going to be pulled from the milestone as it doesn't fit with our "stability" theme. If there is no communication after COB tomorrow, an exception will be required to add it back to the milestone. Please let me or Enhancements Lead @kacole2 know where we stand. Thanks!

verb commented:
@guineveresaenger I have the kubelet code ready, it's just waiting on the API #59416. Let's wait for the API discussion to conclude and remove this from 1.13.

spiffxp commented:
/milestone v1.14
Punting out of v1.13 per discussion above

verb commented:
/remove-lifecycle stale

nikopen commented:
/milestone v1.15

verb committed:
Generated code for Ephemeral Containers in kubelet

verb commented:
/retest

soggiest commented:
Hello! We have begun Code Freeze for 1.15. Is this PR still planned to be included in 1.15? If sop will is be finished in the next week or so?

verb commented:
@soggiest No, this is blocked on API review #59416, which I don't expect to be complete in time for 1.15.

verb commented:
/retest

verb commented:
/retest

verb committed:
Add support for ephemeral containers to the kubelet

verb commented:
/retest

verb commented:
Added release notes

yujuhong commented on pkg/kubelet/container/helpers_test.go:
```diff
@@ -320,6 +324,72 @@ func TestExpandVolumeMountsWithSubpath(t *testing.T) {
+			haveName:      "debug-container",
+			wantContainer: &v1.Container{Name: "debug-container"},
+		},
+	} {
```
Use subtests, i.e., `t.Run()`

verb replied:
Done.

yujuhong commented on pkg/kubelet/container/helpers_test.go:
```diff
@@ -320,6 +324,72 @@ func TestExpandVolumeMountsWithSubpath(t *testing.T) {
+	} {
+		gotContainer := GetContainerSpec(tc.havePod, tc.haveName)
+		if diff := cmp.Diff(tc.wantContainer, gotContainer); diff != "" {
+			t.Errorf("GetContainerSpec for %q returned diff (-want +got):%v", tc.name, diff)
```
Can probably just use `assert` here.

verb replied:
I don't feel strongly, but the output of the manual diff is more readable:
```
        helpers_test.go:390: GetContainerSpec for "ephemeral container" returned diff (-want +got):  &v1.Container{
            -   Name:    "debug-containr",
            +   Name:    "debug-container",
                Image:   "",
                Command: nil,
                ... // 18 identical fields
              }
```

with assert:
```
        helpers_test.go:388:
                Error Trace:    helpers_test.go:388
                Error:          Not equal:
                                expected: &v1.Container{Name:"debug-containr", Image:"", Command:[]string(nil), Args:[]string(nil), WorkingDir:"", Ports:[]v1.ContainerPort(nil),
EnvFrom:[]v1.EnvFromSource(nil), Env:[]v1.EnvVar(nil), Resources:v1.ResourceRequirements{Limits:v1.ResourceList(nil), Requests:v1.ResourceList(nil)}, VolumeMounts:[]v1.VolumeMoun
t(nil), VolumeDevices:[]v1.VolumeDevice(nil), LivenessProbe:(*v1.Probe)(nil), ReadinessProbe:(*v1.Probe)(nil), Lifecycle:(*v1.Lifecycle)(nil), TerminationMessagePath:"", Terminat
ionMessagePolicy:"", ImagePullPolicy:"", SecurityContext:(*v1.SecurityContext)(nil), Stdin:false, StdinOnce:false, TTY:false}
                                actual  : &v1.Container{Name:"debug-container", Image:"", Command:[]string(nil), Args:[]string(nil), WorkingDir:"", Ports:[]v1.ContainerPort(nil),
 EnvFrom:[]v1.EnvFromSource(nil), Env:[]v1.EnvVar(nil), Resources:v1.ResourceRequirements{Limits:v1.ResourceList(nil), Requests:v1.ResourceList(nil)}, VolumeMounts:[]v1.VolumeMou
nt(nil), VolumeDevices:[]v1.VolumeDevice(nil), LivenessProbe:(*v1.Probe)(nil), ReadinessProbe:(*v1.Probe)(nil), Lifecycle:(*v1.Lifecycle)(nil), TerminationMessagePath:"", TerminationMessagePolicy:"", ImagePullPolicy:"", SecurityContext:(*v1.SecurityContext)(nil), Stdin:false, StdinOnce:false, TTY:false}

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,2 +1,2 @@
                                -(*v1.Container)(&Container{Name:debug-containr,Image:,Command:[],Args:[],WorkingDir:,Ports:[],Env:[],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},},VolumeMounts:[],LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:,ImagePullPolicy:,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[],TerminationMessagePolicy:,VolumeDevices:[],})
                                +(*v1.Container)(&Container{Name:debug-container,Image:,Command:[],Args:[],WorkingDir:,Ports:[],Env:[],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},},VolumeMounts:[],LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:,ImagePullPolicy:,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[],TerminationMessagePolicy:,VolumeDevices:[],})

                Test:           TestGetContainerSpec/ephemeral_container
```

yujuhong commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1414,13 +1414,30 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
+			ecSpecs = append(ecSpecs, v1.Container(pod.Spec.EphemeralContainers[i].EphemeralContainerCommon))
+		}
+
+		// TODO(verb): By now we've iterated podStatus 3 times. We could refactor this to make a single
```
File an issue instead.

yujuhong commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -396,17 +397,16 @@ func mountedReadOnlyByPod(podVolume v1.Volume, pod *v1.Pod) bool {
-		if !mountedReadOnlyByContainer(podVolume.Name, &container) {
+
+	mountedReadOnly := true
+	podutil.VisitContainers(&pod.Spec, func(c *v1.Container) bool {
```
cc volumemanager approvers @msau42 @saad-ali @jsafrane

msau42 replied:
Can this be simplified to just `return podutil.VisitContainers(...)`?

verb replied:
I was worried that conflating the two levels of bool returns for "visiting was completed" and "mounted read only by pod" would be confusing, but I tried it out and it makes sense for this "search for a counterexample" sort of pattern.

I've updated to this, thanks for the suggestion.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -610,6 +627,7 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 //  4. Create sandbox if necessary.
 //  5. Create init containers.
 //  6. Create normal containers.
+//  7. Create ephemeral containers.
```
This seems to contradict the comment you had above " Ephemeral containers may be started even if initialization is not yet complete."

If we want to allow the ephemeral containers to be created even when init containers haven't finished running, shouldn't we create them before (5)?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -739,22 +757,40 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
-	if container := podContainerChanges.NextInitContainerToStart; container != nil {
-		// Start the next init container.
+	// Helper containing boilerplate common to starting all types of containers.
+	// typeName is a label used to describe this type of container in log messages.
```
nit: list the types that are currently in use.

verb replied:
Done.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -764,29 +800,14 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
-			klog.V(4).Infof("Backing Off restarting container %+v in pod %v", container, format.Pod(pod))
-			continue
-		}
+		start("container", &pod.Spec.Containers[idx])
```
Why discarding the returned error?

verb replied:
That's the current behavior: start the container, log any errors, move on to the next container. The logging is now done in `start()`

I should mention that the only behavior here that I (intentionally) changed was that the log spam filtering for `ErrImagePullBackOff` was previously only applied to regular containers and not init containers.

yujuhong commented:
/cc @Random-Liu @dashpole

verb committed:
Change order kubelet starts containers

This starts ephemeral containers prior to init containers so that
ephemeral containers will still be started when init containers fail to
start.

Also improves tests and comments with review suggestions.

verb commented:
/retest

msau42 reviewed (commented):
Just one minor comment for volumemanager changes.

/approve

verb committed:
Simplify VisitContainers pattern in volumemanager populator

verb commented:
Cleaning up labels from before api change merged.

/remove-kind api-change
/remove-area test
/remove-area release-eng
/remove-sig apps
/remove-sig auth
/remove-sig api-machinery
/remove-sig release
/remove-sig storage
/remove-sig testing

verb commented:
/kind feature

verb commented:
/retest

verb committed:
Fix returning logs from ephemeral containers

yujuhong commented:
/approve
/lgtm

/retest

verb commented:
/retest
